Adds win_pagefile module

LiranNis opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Adds module that allows management of **Windows** pagefiles.
<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
module: lib/ansible/modules/windows/win_pagefile.ps1
doc: lib/ansible/modules/windows/win_pagefile.py

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
```
ansible 2.3.1.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/etc/ansible/modules/']
  python version = 2.7.5 (default, Aug  2 2016, 04:20:16) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

LiranNis committed:
Adds win_pagefile module

LiranNis committed:
Fixed win_pagefile doc

LiranNis committed:
Fix win_pagefile doc

jhawkesworth commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+
+- name: Query C pagefile
+  win_pagefile:
+    drive=c
```
please can you use key: value rather than key=value for the examples.  Sorry if this sounds fussy but key: value is standard now I believe.

jhawkesworth commented:
Thanks for this.

I have a couple of suggestions (as well as the minor thing about the examples above).

Please could you implement 'check_mode'.  There are now lots of examples as most of the windows modules now support it?

Also, typically modules don't have a 'query' mode.  Instead the usual thing to do is to gather facts.  I don't think there are currently any pagefile facts (I haven't looked), but it would make sense to add them to me.  Obviously that would be a separate pull request.

We are trying to get integration tests for all the windows modules at the moment.  Since you have got a good collection of examples it might not be much more work to create integration tests (which are just playbooks).  Have a look at the test/integration/targets/win* for some examples.  Might make sense to make that into a separate pull request.

Looks good otherwise though, thank you for taking the time to submit this.

LiranNis committed:
Fix win_pagefile doc variable convention

dagwieers commented:
We would like to have integration tests for all (Windows) modules, but in this case I don't know how this will affect our VMs. If it's possible, please add integration tests !

LiranNis committed:
Added check_mode

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,162 @@
+$state = Get-AnsibleParam -obj $params -name "state" -type "str" -default "query" -validateset "present","absent","query"
+$systemManaged = Get-AnsibleParam -obj $params -name "system_managed" -type "bool" -default $false
+
+$ErrorActionPreference = "Stop"
```
Please move this up, to the beginning of the code.
I am also wondering if this is really required now that the powershell.ps1 includes it.
cc: @nitzmahone @jborean @jhawkesworth Anyone ? :-)

LiranNis replied:
I started to write this module when it wasn't part of powershell.ps1 :) I will remove that.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,162 @@
+
+########
+
+$result = @{ 
```
I would move this after the parameter handling section. But that's a personal preference.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+    choices:
+      - true
+      - false
+    default: true
```
So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'yes'
```
And this will do the required to ensure the documentation is consistent.
We prefer 'yes' and 'no' for booleans as it is the most human readable format.
(And we have to make it strings, otherwise YAML will make it True/False again)

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+    choices:
+      - true
+      - false
+    default: false
```
So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'no'
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+      - Configures AutomaticManagedPagefile for the entire system.
+    choices:
+      - true
+      - false
```
So booleans require the following statements instead if the above:
```yaml
  type: bool
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+    choices:
+      - true
+      - false
+    default: false
```
So booleans require the following statements instead if the above:
```yaml
  type: bool
  default: 'no'
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+- InitialSize 0 and MaximumSize 0 means the pagefile is managed by the system.
+- Value out of range exception may be caused by several different issues, two common problems - No such drive, Pagefile size is too small.
+- Setting a pagefile when AutomaticManagedPagefile is on will disable the AutomaticManagedPagefile.
+author: "Liran Nisanov (@LiranNis)"
```
Please turn this into a list, without the quotes (as we tend to only add quotes when they are required by YAML).
So it should become:
```yaml
author:
- Liran Nisanov (@LiranNis)
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,156 @@
+  drive:
+    description:
+      - The drive of the pagefile.
+    required: false
```
If it is not required, you don't have to add `required: false`. This is implicit (and makes it easier to see what is required).

LiranNis committed:
Changed win_pagefile module&doc to the convention

LiranNis committed:
added win_pagefile integration tests

LiranNis commented:
@jhawkesworth, I added integration tests and check_mode. 
About the query part, after it will be added to the facts (I have few things to add to there) we can remove the query and return pagefiles for every type of task, as part of the result variable, it will require to change the integration tests as well.

The virtual machines of the integration tests have only C drive? I wrote it to check only one pagefile over C drive.
Another question about the vms, you create a new vm (or revert to snapshot) for every run of the tests? I need to know if I need to return the pagefiles to automatic if you don't do so (someone wrote comment about that and deleted it, so I am not sure).

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,160 @@
+
+Function Remove-Pagefile($path) 
+{
+    Get-WmiObject Win32_PageFileSetting | WHERE { $_.Name -eq $path } | Remove-WmiObject
```
Could you add here -WhatIf support ?

```powershell
Function Remove-Pagefile($path, $whatif) {
    Get-WmiObject Win32_PageFileSetting | WHERE { $_.Name -eq $path } | Remove-WmiObject -WhatIf:$whatif
}
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,160 @@
+    $currentPageFiles = Get-WmiObject Win32_PageFileSetting
+    if ($currentPageFiles -ne $null) {
+        if (-not $check_mode) {
+            $currentPageFiles | Remove-WmiObject | Out-Null
```
I think you could do here:

```powershell
$currentPageFiles | Remove-WmiObject -WhatIf:$check_mode | Out-Null
```

And avoid the conditional.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,160 @@
+    {
+        try {
+            if (-not $check_mode) {
+                Remove-Pagefile $fullPath
```
This would become:
```powershell
Remove-Pagefile $fullPath -WhatIf:$check_mode
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,160 @@
+        {
+            try {
+                if (-not $check_mode) {
+                    Remove-Pagefile $fullPath
```
This would become:
```powershell
Remove-Pagefile $fullPath -WhatIf:$check_mode
```

dagwieers reviewed (changes requested):
Good work !

jhawkesworth commented:
@LiranNis - right now the integration tests run on amazon - I don't know the details of the AMIs used but I guess it is safest to assume only a C: drive.  

I think safest thing to do is restore the page file state to whatever it was before the integration tests started.  While this might be slower to run, it means the tests can be re-run without having to reset the machine state.

LiranNis committed:
Changed check_mode blocks to whatif, fixed a bug

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -114,11 +108,11 @@ if ($state -eq "absent") {
-                $pagefile.MaximumSize = $maximumSize
-                if (-not $check_mode) {
+            if (-not $check_mode) {
+                $pagefile = Set-WmiInstance -Class Win32_PageFileSetting -Arguments @{name = $fullPath; InitialSize = 0; MaximumSize = 0}
```
Forgive me for being pedantic, but here again I would prefer a `-WhatIf:$check_mode` for Set-WmiInstance.
Obviously the rest of the block does still need the `if (-not $check_mode)`.

The reason why we prefer the `-WhatIf` functionality is that it is likely that the function also does more than just skip it based on the boolean. We expect that where possible the function also verifies other conditions within the call.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,148 @@
+version_added: "2.4"
+short_description: Query or change pagefile configuration
+description:
+    - Query current pagefile configuration
```
And just because we show signs of OCD. The items in `description:` should end with a dot.
This is not true for `short_description:`, there we do not allow a dot.

Isn't life beautiful like that ? :-P

dagwieers commented on test/integration/targets/win_pagefile/tasks/main.yml:
```diff
@@ -0,0 +1,136 @@
+  set_fact: 
+    expected:
+      pagefiles: 
+      - { caption: "C:\\ 'pagefile.sys'",
```
Just a FYI, this can be pure YAML too:
```yaml
- name: Set fact for pagefile expected result
  set_fact:
    expected:
      pagefiles:
      - caption: C:\ 'pagefile.sys'
        description: "'pagefile.sys' @ C:\\"
        initial_size: 1024
        maximum_size: 2048
        name: C:\pagefile.sys
```

LiranNis replied:
Thanks, I didn't know that, it's a lot easier :)

LiranNis committed:
Added whatif to set-wmiinstance, changed docs

Added whatif in Set-WMIInstance
Added dots to end of decription lines

LiranNis committed:
Returns to original state at the end, more tests

Added override and no override integration tests
Pagefiles now return to same state as before at the end of the integration test

LiranNis committed:
Remove extra line

LiranNis commented:
To debug the first error I am downloading and installing Windows Server 2008 SP2 right now.

The second error caused by a weird type of drive, not sure what type is Z: drive, but you can't use Test-Path on it, I can remove this condition but it may cause setting up pagefiles on CD drives for example.

LiranNis committed:
Added test_path var to win_pagefile

LiranNis committed:
Set test_path as 'no' in integration

mattclay commented:
CI failure in the `win_pagefile` test on 2008 SP2 for task `Return all previous pagefiles settings`:
```
{
"changed": true, 
"failed": true, 
"msg": "All items completed"
}
```

jborean93 commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+            $result.changed = $true
+        }
+    } catch {
+        Fail-Json $result "Failed to set AutomaticManagedPagefile $_.Exception.Message"
```
You need to enclose the exception part in brackets like `$($_.Exception.Message)`

jborean93 commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+            Remove-Pagefile $fullPath -whatif:$check_mode
+            $result.changed = $true
+        } catch {
+            Fail-Json $result "Failed to remove pagefile $_.Exception.Message"
```
You need to enclose the exception part in brackets like `$($_.Exception.Message)`

jborean93 commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+        # Get automatic managed pagefile state
+        $result.automatic_managed_pagefiles = (Get-WmiObject -Class win32_computersystem).AutomaticManagedPagefile
+    } catch {
+        Fail-Json $result "Failed to query current pagefiles $_.Exception.Message"
```
You need to enclose the exception part in brackets like `$($_.Exception.Message)`

jborean93 commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+                Remove-Pagefile $fullPath -whatif:$check_mode
+                $result.changed = $true
+            } catch {
+                Fail-Json $result "Failed to remove current pagefile $_.Exception.Message"
```
You need to enclose the exception part in brackets like `$($_.Exception.Message)`

jborean93 commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+            $result.changed = $true
+        }
+    } catch {
+        Fail-Json $result "Failed to set pagefile $_.Exception.Message"
```
You need to enclose the exception part in brackets like `$($_.Exception.Message)`

jborean93 commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,153 @@
+      - The drive of the pagefile.
+  initial_size:
+    description:
+      - The initial size of the pagefile.
```
Are you able to specify whether this is in bytes, kilobytes, megabytes etc?

jborean93 commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,153 @@
+      - The initial size of the pagefile.
+  maximum_size:
+    description:
+      - The maximum size of the pagefile.
```
Are you able to specify whether this is in bytes, kilobytes, megabytes etc?

jborean93 reviewed (changes requested):
Looks good, thanks for the contribution @LiranNis, I've just added a few comments.

LiranNis committed:
Added unit to docs and enclosed exception message

LiranNis commented:
@jborean93 thanks for the comments, fixed it.
I am still trying to understand why there is an error in the 2008 SP2 server.

jhawkesworth commented:
That's an odd one.  Looks like others have hit this:

https://www.vistax64.com/powershell/189685-adding-new-pagefile-another-drive-than-system-drive.html

LiranNis commented:
From what I saw the VMs don't have Z drive, I added a work around to return it to the way it was (test_path variable).
In the case of the 2008 SP2 server there is not enough free space in C drive so the pagefile can't be set as it was before.
Reference: https://social.technet.microsoft.com/Forums/windows/en-US/46acf090-1da0-42b2-8a37-5628d7fc8068/disabled-automatic-pagefile-management-now-have-pagefile-on-every-volume?forum=winservercore

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+            $result.changed = $true
+        }
+    } catch {
+        Fail-Json $result "Failed to set pagefile $($_.Exception.Message)"
```
This try-catch block is rather large.
I would put 2 different try-catch block around `Set-WmiInstance` and around `$pagefile.Put()`.
Everything else shouldn't fail, and we get more granular error-handling.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,153 @@
+        # Get automatic managed pagefile state
+        $result.automatic_managed_pagefiles = (Get-WmiObject -Class win32_computersystem).AutomaticManagedPagefile
+    } catch {
+        Fail-Json $result "Failed to query current pagefiles $($_.Exception.Message)"
```
Same here, try to use try-catch for as few statements as possible.

LiranNis committed:
More granular try-catch blocks

LiranNis commented:
@dagwieers thanks for the advice, I hope it's better now.

dagwieers commented:
@LiranNis Perfect ! Much better. We don't have the tendency to catch only specific exceptions in the Windows world (i.e. I haven't seen any Windows module do this correctly), so I am not going to request that. (It's also not always obvious in what ways an instruction can fail, so...)

dagwieers reviewed (approved):
LGTM

dagwieers commented:
+label new_module

LiranNis committed:
Added workaround to avoid value out of range

LiranNis commented:
@jhawkesworth thanks! I decided to use the trick in your link as a workaround, because I still prefer to use WMI functions if I can. 
I tried other ways, most of them caused WMI Provider to crash, so that is the only way.
Microsoft doesn't allow to set these pagefile values through WMI (at least in 2008 SP2), but there is no check through the registry.

Do you think this workaround is ok? Maybe there needs to be a flag to allow users that don't want to use this workaround and prefer fail?
BTW, Can someone please check why there is a pagefile on drive Z when there is no drive Z on the server?

LiranNis committed:
Deleted wrong line ending

LiranNis commented:
@jborean93 I fixed the changes you asked for, can you review it?

dagwieers commented:
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

LiranNis commented:
@dagwieers in my opinion the module is finished, I am waiting for more reviews.

dagwieers commented on lib/ansible/modules/windows/win_pagefile.ps1:
```diff
@@ -0,0 +1,183 @@
+#
+# Copyright 2017, Liran Nisanov <lirannis@gmail.com>
+#
+# Ansible is free software: you can redistribute it and/or modify
```
So, quite recently there was a change to how we add the GPL license to each individual file. It was changed to only add the following one-liner:
```powershell
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

dagwieers commented on lib/ansible/modules/windows/win_pagefile.py:
```diff
@@ -0,0 +1,153 @@
+
+# (c) 2017, Liran Nisanov <lirannis@gmail.com>
+#
+# This file is part of Ansible
```
Same here, the only line remaining (after the Copyright statement(s) is:
```yaml
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

dagwieers commented:
@LiranNis Reading back over the discussions, I think it would be best to remove the query-information from the module. And move those bits in the default Windows facts. It is easier to leave it out, then to remove it later (as we have to deprecate it for 4 versions...).

@nitzmahone @jborean93 @jhawkesworth So this is about to be merged. Any last remarks ?

jborean93 reviewed (approved):
I'm good to merge this and get it in before the 2.4 freeze.

LiranNis committed:
Changed license to one-line

LiranNis committed:
Removed space in line ending

LiranNis committed:
Try to fix python2.6 error

LiranNis commented:
@dagwieers I changed the license, if I will remove the query part I won't be able to get the current pagefiles and return it to the same values at the end of the integration run,
And I did few more checks in the integration with it, which means the integration will only check changed / ok / failed and not the values itself, when it will be in facts that is probably what I will do.
I still want to add pagefiles and disks to the facts, I will work on it soon.

LiranNis committed:
Try 2 to fix python2.6 error

LiranNis committed:
Add separating line again

genadipost commented:
@dagwieers This module can be very useful for me (and many others). What holds this merge request from getting in before 2.4 freeze? (and how can i help)

dagwieers commented:
@genadipost I expect that most open Windows module PRs will be merged before 29/8 (the v2.4 freeze). I don't see a reason to not merge this, but leave the final decision and merge to @jhawkesworth @jborean93 @nitzmahone . Who have a better trackrecord with Windows than me ;-)

dagwieers reviewed (approved):
LGTM

LiranNis commented:
@jhawkesworth @jborean93 @nitzmahone is it going to be merged in v2.4?

jborean93 commented:
Let's get this in before the freeze, we can deal with bugs in that phase if any are encountered.
